### PR TITLE
trezor: support c++17 and protobuf v25, libusb fix #smol

### PR DESCRIPTION
**Pull request by ph4r05**
- fixes support for C++ 17 && Protobuf v22+
- fixes libusb compile check for ubuntu

@selsta c++17 support came sooner than expected 😅 I've modified the code a bit to protobufv25 and c++17 works

### CODE REVIEW

**Comment by ph4r05** [comment]
@selsta c++17 support came sooner than expected 😅 I've modified the code a bit to protobufv25 and c++17 works

**Comment by ph4r05** [comment]
@jeffro256 @moneromooo-monero when you have a minute for a review, this one is small 😅  thanks!

**Comment by ph4r05** [comment]
Oor maybe @vtnerd @rbrunner7 ? :)

This PR is minimalistic, thanks!

**Comment by selsta** [comment]
I will review it but didn't get to it yet.

File: cmake/CheckTrezor.cmake [hunk 1]
```diff
@@ -150,9 +156,14 @@ if(Protobuf_FOUND AND USE_DEVICE_TREZOR AND TREZOR_PYTHON)
     endif()
 
     if(USE_DEVICE_TREZOR_PROTOBUF_TEST)
-        # For now, Protobuf v21 is the maximum supported version as v23 requires C++17. TODO: Remove once we move to C++17
-        if(${Protobuf_VERSION} GREATER 21)
-            trezor_fatal_msg("Trezor: Unsupported Protobuf version ${Protobuf_VERSION}. Please, use Protobuf v21.")
+        if(${CMAKE_CXX_STANDARD} LESS 17 AND ${Protobuf_VERSION} GREATER 21)
```
Comment (vtnerd) [inline cmake/CheckTrezor.cmake:159]: Isn't this a duplicate from above?

Comment (ph4r05) [reply cmake/CheckTrezor.cmake]: yes, it is a duplicate, removed now

**Comment by ph4r05** [comment]
Hello, just checking if you have a minute for the review :)

**Comment by jeffro256** [comment]
I'm getting linking errors with target `release-static` on Ubuntu 23. 

Selection from error messages:
```
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_signal_new':
(.text+0xbb): undefined reference to `event_set'
/usr/bin/ld: (.text+0xc7): undefined reference to `event_base_set'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_event_new':
(.text+0x15d): undefined reference to `event_set'
/usr/bin/ld: (.text+0x16a): undefined reference to `event_base_set'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_timer_add':
(.text+0x1e6): undefined reference to `event_set'
/usr/bin/ld: (.text+0x1f2): undefined reference to `event_base_set'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_event_base_free':
(.text+0x25d): undefined reference to `event_base_free'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `ub_default_event_base':
(.text+0x2cd): undefined reference to `event_base_new'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `ub_get_event_sys':
(.text+0x3a6): undefined reference to `event_get_version'
/usr/bin/ld: (.text+0x3bc): undefined reference to `event_base_get_method'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_event_del':
(.text+0x1b9): undefined reference to `event_del'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_timer_add':
(.text+0x205): undefined reference to `event_add'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_event_add':
(.text+0x229): undefined reference to `event_add'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_event_base_loopexit':
(.text+0x239): undefined reference to `event_base_loopexit'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_event_base_dispatch':
(.text+0x249): undefined reference to `event_base_dispatch'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_signal_del':
(.text+0x279): undefined reference to `event_del'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_timer_del':
(.text+0x289): undefined reference to `event_del'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(ub_event_pluggable.o): in function `my_signal_add':
(.text+0x299): undefined reference to `event_add'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(random.o): in function `ub_initstate':
(.text+0x40): undefined reference to `nettle_yarrow256_init'
/usr/bin/ld: (.text+0x62): undefined reference to `nettle_yarrow256_seed'
/usr/bin/ld: (.text+0x6a): undefined reference to `nettle_yarrow256_is_seeded'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(random.o): in function `ub_random':
(.text+0x139): undefined reference to `nettle_yarrow256_random'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(random.o): in function `ub_random_max':
(.text+0x203): undefined reference to `nettle_yarrow256_random'
/usr/bin/ld: (.text+0x251): undefined reference to `nettle_yarrow256_random'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `_digest_nettle':
(.text+0x49): undefined reference to `nettle_sha256_init'
/usr/bin/ld: (.text+0x57): undefined reference to `nettle_sha256_update'
/usr/bin/ld: (.text+0x67): undefined reference to `nettle_sha256_digest'
/usr/bin/ld: (.text+0xa6): undefined reference to `nettle_sha512_init'
/usr/bin/ld: (.text+0xb4): undefined reference to `nettle_sha512_update'
/usr/bin/ld: (.text+0xc4): undefined reference to `nettle_sha512_digest'
/usr/bin/ld: (.text+0xd7): undefined reference to `nettle_sha384_init'
/usr/bin/ld: (.text+0xe5): undefined reference to `nettle_sha512_update'
/usr/bin/ld: (.text+0xf5): undefined reference to `nettle_sha384_digest'
/usr/bin/ld: (.text+0x107): undefined reference to `nettle_sha1_init'
/usr/bin/ld: (.text+0x115): undefined reference to `nettle_sha1_update'
/usr/bin/ld: (.text+0x125): undefined reference to `nettle_sha1_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `_verify_nettle_rsa':
(.text+0x1cf): undefined reference to `nettle_rsa_public_key_init'
/usr/bin/ld: (.text+0x1f9): undefined reference to `nettle_mpz_set_str_256_u'
/usr/bin/ld: (.text+0x211): undefined reference to `nettle_mpz_set_str_256_u'
/usr/bin/ld: (.text+0x21f): undefined reference to `nettle_mpz_init_set_str_256_u'
/usr/bin/ld: (.text+0x240): undefined reference to `nettle_rsa_public_key_clear'
/usr/bin/ld: (.text+0x248): undefined reference to `__gmpz_clear'
/usr/bin/ld: (.text+0x2cb): undefined reference to `nettle_sha1_init'
/usr/bin/ld: (.text+0x2d9): undefined reference to `nettle_sha1_update'
/usr/bin/ld: (.text+0x2f1): undefined reference to `nettle_sha1_digest'
/usr/bin/ld: (.text+0x301): undefined reference to `nettle_rsa_sha1_verify_digest'
/usr/bin/ld: (.text+0x310): undefined reference to `nettle_rsa_public_key_clear'
/usr/bin/ld: (.text+0x318): undefined reference to `__gmpz_clear'
/usr/bin/ld: (.text+0x343): undefined reference to `nettle_sha512_init'
/usr/bin/ld: (.text+0x351): undefined reference to `nettle_sha512_update'
/usr/bin/ld: (.text+0x369): undefined reference to `nettle_sha512_digest'
/usr/bin/ld: (.text+0x379): undefined reference to `nettle_rsa_sha512_verify_digest'
/usr/bin/ld: (.text+0x39b): undefined reference to `nettle_sha256_init'
/usr/bin/ld: (.text+0x3a9): undefined reference to `nettle_sha256_update'
/usr/bin/ld: (.text+0x3c1): undefined reference to `nettle_sha256_digest'
/usr/bin/ld: (.text+0x3d1): undefined reference to `nettle_rsa_sha256_verify_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `_verify_nettle_ecdsa':
(.text+0x466): undefined reference to `nettle_dsa_signature_init'
/usr/bin/ld: (.text+0x484): undefined reference to `nettle_get_secp_384r1'
/usr/bin/ld: (.text+0x499): undefined reference to `nettle_ecc_point_init'
/usr/bin/ld: (.text+0x4a9): undefined reference to `nettle_mpz_init_set_str_256_u'
/usr/bin/ld: (.text+0x4bf): undefined reference to `nettle_mpz_init_set_str_256_u'
/usr/bin/ld: (.text+0x4cf): undefined reference to `nettle_mpz_set_str_256_u'
/usr/bin/ld: (.text+0x4e2): undefined reference to `nettle_mpz_set_str_256_u'
/usr/bin/ld: (.text+0x4fc): undefined reference to `nettle_sha384_init'
/usr/bin/ld: (.text+0x514): undefined reference to `nettle_sha512_update'
/usr/bin/ld: (.text+0x524): undefined reference to `nettle_sha384_digest'
/usr/bin/ld: (.text+0x532): undefined reference to `nettle_ecc_point_set'
/usr/bin/ld: (.text+0x547): undefined reference to `nettle_ecdsa_verify'
/usr/bin/ld: (.text+0x551): undefined reference to `__gmpz_clear'
/usr/bin/ld: (.text+0x55c): undefined reference to `__gmpz_clear'
/usr/bin/ld: (.text+0x564): undefined reference to `nettle_ecc_point_clear'
/usr/bin/ld: (.text+0x56c): undefined reference to `nettle_dsa_signature_clear'
/usr/bin/ld: (.text+0x591): undefined reference to `nettle_get_secp_256r1'
/usr/bin/ld: (.text+0x5a6): undefined reference to `nettle_ecc_point_init'
/usr/bin/ld: (.text+0x5b6): undefined reference to `nettle_mpz_init_set_str_256_u'
/usr/bin/ld: (.text+0x5cc): undefined reference to `nettle_mpz_init_set_str_256_u'
/usr/bin/ld: (.text+0x5dc): undefined reference to `nettle_mpz_set_str_256_u'
/usr/bin/ld: (.text+0x5ef): undefined reference to `nettle_mpz_set_str_256_u'
/usr/bin/ld: (.text+0x609): undefined reference to `nettle_sha256_init'
/usr/bin/ld: (.text+0x621): undefined reference to `nettle_sha256_update'
/usr/bin/ld: (.text+0x631): undefined reference to `nettle_sha256_digest'
/usr/bin/ld: (.text+0x63f): undefined reference to `nettle_ecc_point_set'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `secalgo_hash_create_sha384':
(.text+0x6e7): undefined reference to `nettle_sha384_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `secalgo_hash_create_sha512':
(.text+0x71a): undefined reference to `nettle_sha512_init'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `secalgo_hash_update':
(.text+0x76b): undefined reference to `nettle_sha512_update'
/usr/bin/ld: (.text+0x788): undefined reference to `nettle_sha512_update'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `secalgo_hash_final':
(.text+0x7c5): undefined reference to `nettle_sha512_digest'
/usr/bin/ld: (.text+0x80a): undefined reference to `nettle_sha384_digest'
/usr/bin/ld: /usr/lib/x86_64-linux-gnu/libunbound.a(val_secalgo.o): in function `verify_canonrrset':
(.text+0xa7a): undefined reference to `nettle_ed25519_sha512_verify'
collect2: error: ld returned 1 exit status

```

**Comment by ph4r05** [comment]
@jeffro256 it does not seem related to this PR, can you pls try again?

**Comment by jeffro256** [comment]
Pulled latest commit and retried compiling, got the same results. I can run `make release-static` just fine on master. I'm not quite sure what the reason is, I'll keep looking...

**Comment by ph4r05** [comment]
looks similar to https://github.com/monero-project/monero/issues/8439

**Comment by ph4r05** [comment]
@jeffro256 It looks completely unrelated, static linking for libunbound requires also libnettle, which is not linked. The following commit fixed the linking on Ubuntu 23.04 https://github.com/monero-project/monero/pull/9064/commits/04ccc0ff8e63cf7bac8698f7f96008b74b4bbef1

Now checking if it passes on the CI

**Comment by ph4r05** [comment]
@jeffro256 ok I figured out that libunbound + nettle does not statically compile on Ubuntu 23.04 by default. I tried several combinations, nothing worked. Default nettle did not have for example `nettle_mpz_set_str_256_uz` required by unbound.

After installing unbound-1.61.1 according to https://github.com/sethforprivacy/simple-monerod-docker/pull/61 compilation started to work out of the box.

```shell
wget https://www.nlnetlabs.nl/downloads/unbound/unbound-1.16.1.tar.gz && \
    echo "2fe4762abccd564a0738d5d502f57ead273e681e92d50d7fba32d11103174e9a  unbound-1.16.1.tar.gz" | sha256sum -c && \
    tar -xzf unbound-1.16.1.tar.gz && \
    rm unbound-1.16.1.tar.gz && \
    cd unbound-1.16.1 && \
    ./configure --disable-shared --enable-static --without-pyunbound --with-libexpat=/usr --with-ssl=/usr --with-libevent=no --without-pythonmodule --disable-flto --with-pthreads --with-libunbound-only --with-pic && \
    make -j${NPROC:-$(nproc)} && \
    make -j${NPROC:-$(nproc)} install
```

**Comment by ph4r05** [comment]
@selsta would you pls find a minute for a review? :)

**Comment by ph4r05** [comment]
ooor @vtnerd 🙏 ? :)

**Comment by ph4r05** [comment]
ooor @vtnerd 🙏 ? @moneromooo-monero 🙏 ? :)

**Comment by selsta** [comment]
I will try to review it next week

File: src/device_trezor/README.md [hunk 2]
```diff
@@ -15,12 +15,22 @@ Please, refer to [monero readme](https://github.com/trezor/trezor-firmware/blob/
 
 ## Dependencies
 
-Trezor uses [Protobuf](https://protobuf.dev/) library. As Monero is compiled with C++14, the newest Protobuf library version cannot be compiled because it requires C++17 (through its dependency Abseil library).
-This can result in a compilation failure.
+Trezor uses [Protobuf](https://protobuf.dev/) library. Monero is now compiled with C++ 17 by default.
+Thus, protobuf v23 and higher has to be used.
```
Comment (selsta) [inline src/device_trezor/README.md:19]: Is v23 a necessity? We build release binaries with v21 (https://github.com/monero-project/monero/blob/master/contrib/depends/packages/native_protobuf.mk#L2) and it compiled with Trezor support enabled according to CI in this PR.

File: cmake/CheckTrezor.cmake [hunk 3]
```diff
@@ -75,19 +75,25 @@ if (USE_DEVICE_TREZOR)
...
     if (NOT Protobuf_FOUND)
         FIND_PACKAGE(Protobuf)
     endif()
 
     _trezor_protobuf_fix_vars()
 
     # Early fail for optional Trezor support
-    if(${Protobuf_VERSION} GREATER 21)
-        trezor_fatal_msg("Trezor: Unsupported Protobuf version ${Protobuf_VERSION}. Please, use Protobuf v21.")
+    if(${CMAKE_CXX_STANDARD} LESS 17 AND ${Protobuf_VERSION} GREATER 21)
```
Comment (selsta) [inline cmake/CheckTrezor.cmake:92]: If there is no protobuf version installed, this `if` fails with this CMake error

```
CMake Error at cmake/CheckTrezor.cmake:92 (if):
  if given arguments:

    "17" "LESS" "17" "AND" "GREATER" "21"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:704 (include)


-- Configuring incomplete, errors occurred!
```

Which aborts CMake. This is unintended behaviour for `USE_DEVICE_TREZOR=ON` and `USE_DEVICE_TREZOR_MANDATORY=OFF`. The issue already exists in the current code (see https://github.com/monero-project/monero/issues/9040) but it can be fixed together with this PR.

File: cmake/CheckTrezor.cmake [hunk 4]
```diff
@@ -75,19 +75,25 @@ if (USE_DEVICE_TREZOR)
     # Protobuf is required to build protobuf messages for Trezor
     include(FindProtobuf OPTIONAL)
 
-    FIND_PACKAGE(Protobuf CONFIG)
+    # PkgConfig works better with new Protobuf
+    find_package(PkgConfig REQUIRED)
```
Comment (selsta) [inline cmake/CheckTrezor.cmake:79]:
```suggestion
    find_package(PkgConfig QUIET)
```

at least that's what we use in `cmake/FindSodium.cmake`, is required here necessarsy?

File: cmake/CheckTrezor.cmake (hunk 3 above)
Comment (ph4r05) [reply cmake/CheckTrezor.cmake]: aah you are right, I exchanged those two ifs so first protobuf is checked. it fixes it.

File: src/device_trezor/README.md (hunk 2 above)
Comment (ph4r05) [reply src/device_trezor/README.md]: No, it is not a hard requirement, it is changed now

**Comment by ph4r05** [comment]
@selsta thanks for the review! pls when you have a minute to check again, I addressed your comments. Thanks a lot!

File: src/device_trezor/README.md [hunk 5]
```diff
@@ -15,19 +15,29 @@ Please, refer to [monero readme](https://github.com/trezor/trezor-firmware/blob/
 
 ## Dependencies
 
-Trezor uses [Protobuf](https://protobuf.dev/) library. As Monero is compiled with C++14, the newest Protobuf library version cannot be compiled because it requires C++17 (through its dependency Abseil library).
-This can result in a compilation failure.
+Trezor uses [Protobuf](https://protobuf.dev/) library. Monero is now compiled with C++ 17 by default.
+Protobuf v21+ is required. Note that Protobuf v23+ requires C++ 17.
```
Comment (selsta) [inline src/device_trezor/README.md:19]: Why is v21 required?

Comment (ph4r05) [reply src/device_trezor/README.md:19]: v21 (and greater) is the tested version. Would you prefer different wording?

Comment (selsta) [reply src/device_trezor/README.md:19]: Yes, maybe reword it as something lime v21+ is tested, older versions are not guaranteed to work.

Comment (ph4r05) [reply src/device_trezor/README.md]: done :)

**Comment by ph4r05** [comment]
@selsta when you have a minute, all comments should be addressed by now. thanks! :)

**Comment by ph4r05** [comment]
@selsta ping ping :)

**Comment by selsta** [comment]
@ph4r05 sorry, was busy with other things. please squash and then it can be merged

**Comment by ph4r05** [comment]
squashed @selsta, thank you!

**Review by selsta — APPROVED** [review]

**Review by jeffro256 — APPROVED** [review]
